dask/array/utils.py

```diff
@@ -388,6 +393,36 @@ def zeros_like_safe(a, shape, **kwargs):
         return np.zeros(shape, **kwargs)
 
 
+def _array_like_safe(np_func, da_func, a, like, **kwargs):
+    if like is a and hasattr(a, "__array_function__"):
+        return a
+
+    if isinstance(like, Array):
+        return da_func(a, **kwargs)
+    elif isinstance(a, Array):
+        if _is_cupy_type(a._meta):
+            a = a.compute(scheduler="sync")
```

Comment:
I'm curious why we're using the synchronous scheduler here. I would have expected us to dispatch to the default scheduler. This is what we do in other situations where we need to trigger a compute, like computing `len` of a Dask DataFrame 

https://github.com/dask/dask/blob/2632bbce82d471b7fef426729faec9a4e78ee137/dask/dataframe/core.py#L550-L553

Reply:
I had to dig for some old comments, because I remember having a similar conversation with @mrocklin back in the early days of `__array_function__` support. Please see https://github.com/dask/dask/pull/4543#discussion_r270291526 . This was the case for which Matt suggested we use the synchronous scheduler in the end to compute `_meta`, but maybe I'm ignorant of this situation and perhaps we really don't need the sync scheduler here, given we're computing the actual CuPy data array, and not just an empty one. I'm happy to do it either way, I'm just not confident I can make the best decision on my own for this particular use case.

Reply:
Thanks for the additional context @pentschev. Given that we're, as you mentioned, computing a full array and not just a small `_meta` array this use case seems more in line with other implicit `compute`s like `len(ddf)` and `ddf.to_parquet(...)` where we use the default scheduler. As a user, if I was using a distributed cluster, I would be surprised if a large implicit `compute` was triggered and it didn't use the cluster. @jakirkham do you have thoughts on this topic?

Reply:
Yeah for context Peter and I discussed this a bit upthread ( https://github.com/dask/dask/pull/7162#discussion_r569769319 ).

Generally I must confess I still don't think using `.compute` is ideal (this can have high overhead as Matt also notes in the thread Peter referenced), but haven't been able to come up with a better solution atm. Peter has also since limited this to CuPy arrays, which avoids affecting other cases. 

To the point of `"sync"` specifically, I think in the cases where we are using this function above (NumPy/CuPy arrays, empty `list`s, scalars, etc.) this seems like the right choice as they are small in-memory objects that we are merely testing out to find the expected result. So using `"sync"` seems like the right choice, but I could be overlooking something.

Personally I think we should flag this use of `.compute` in an issue and rework this after the release to avoid it (if at all possible). The bug it fixes is unfortunately pretty critical for NumPy 1.20 support. Anyways this is just my opinion. Feel free to push back

Reply:
> Generally I must confess I still don't think using .compute is ideal (this can have high overhead as Matt also notes in the thread Peter referenced), but haven't been able to come up with a better solution atm. Peter has also since limited this to CuPy arrays, which avoids affecting other cases.

Agreed. I like your idea of flagging this as an issue and trying to improve things where possible after the release.

> this seems like the right choice as they are small in-memory objects that we are merely testing out to find the expected result

That's a good point. The only case I could see where we might trigger a `compute` on a potentially large Dask array is when the first two arguments of `da.percentile` are both Dask arrays backed by CuPy arrays. That said, I'm happy to  merge as-is and revisit post-release

Reply:
Filed issue ( https://github.com/dask/dask/issues/7181 ) on revisiting the `compute` call